File: go/capstone-message-queue/protocol/protocol.go
// Package protocol defines the wire types for the message queue's
// newline-delimited JSON over TCP protocol.
//
// Every message has a "type" field. Payloads and keys are standard base64.
//
// Request types (client → broker):
//
//	{"type":"create_topic","topic":"events","partitions":3}
//	{"type":"produce","topic":"events","key":"dXNlcjE=","payload":"aGVsbG8="}
//	{"type":"fetch","topic":"events","partition":0,"offset":5}
//	{"type":"fetch_batch","topic":"events","partition":0,"offset":5,"max_count":100}
//	{"type":"metadata"}
//	{"type":"join_group","group":"g","topics":["events"]}
//	{"type":"leave_group","group":"g","member_id":"member-0"}
//	{"type":"commit_offset","group":"g","topic":"events","partition":0,"offset":42}
//	{"type":"fetch_offset","group":"g","topic":"events","partition":0}
package protocol

// Request is a decoded client request.
// Payload and Key are base64 strings in the wire format.
type Request struct {
	Type       string   `json:"type"`
	Topic      string   `json:"topic,omitempty"`
	Topics     []string `json:"topics,omitempty"`
	Partitions int      `json:"partitions,omitempty"`
	Partition  uint32   `json:"partition,omitempty"`
	Offset     uint64   `json:"offset,omitempty"`
	MaxCount   int      `json:"max_count,omitempty"`
	Key        string   `json:"key,omitempty"`     // base64-encoded bytes
	Payload    string   `json:"payload,omitempty"` // base64-encoded bytes
	Group      string   `json:"group,omitempty"`
	MemberID   string   `json:"member_id,omitempty"`
}

// TopicMeta describes a single topic in a metadata response.
type TopicMeta struct {
	Name       string `json:"name"`
	Partitions int    `json:"partitions"`
}

// Assignment is a (topic, partition) pair assigned to a consumer group member.
type Assignment struct {
	Topic     string `json:"topic"`
	Partition uint32 `json:"partition"`
}

File: rust/capstone-message-queue/src/protocol.rs
//! Wire protocol: newline-delimited JSON over TCP.
//!
//! **Payload encoding:** raw bytes are base64-encoded inside JSON strings so
//! the protocol stays human-readable and debuggable with `nc` / `telnet`.
//!
//! # Request types
//!
//! ```json
//! {"type":"create_topic","topic":"events","partitions":3}
//! {"type":"produce","topic":"events","payload":"aGVsbG8="}
//! {"type":"produce","topic":"events","key":"dXNlcjE=","payload":"aGVsbG8="}
//! {"type":"fetch","topic":"events","partition":0,"offset":5}
//! {"type":"fetch_batch","topic":"events","partition":0,"offset":5,"max_count":100}
//! {"type":"metadata"}
//! ```
//!
//! # Response types
//!
//! ```json
//! {"type":"topic_created","topic":"events","partitions":3}
//! {"type":"produced","partition":0,"offset":42}
//! {"type":"record","offset":5,"payload":"aGVsbG8="}
//! {"type":"end"}
//! {"type":"metadata","topics":[{"name":"events","partitions":3}]}
//! {"type":"error","message":"topic not found: orders"}
//! ```

use serde::{Deserialize, Serialize};

// ── Request ───────────────────────────────────────────────────────────────────

#[derive(Debug, Deserialize)]
#[serde(tag = "type", rename_all = "snake_case")]
pub enum Request {
    /// Create a topic with `partitions` partitions. Idempotent.
    CreateTopic { topic: String, partitions: u32 },

    /// Append `payload` (base64) to `topic`.
    /// Optional `key` (base64) determines the partition via FNV-1a hash;
    /// omitting `key` uses round-robin.
    Produce {
        topic: String,
        #[serde(default)]
        key: Option<String>,
        payload: String,
    },

    /// Read the single record at `offset` from `topic`/`partition`.
    Fetch {
        topic: String,
        partition: u32,
        offset: u64,
    },

    /// Read up to `max_count` records from `topic`/`partition` starting at
    /// `offset`. Responds with zero or more `record` messages then `end`.
    FetchBatch {
        topic: String,
        partition: u32,
        offset: u64,
        max_count: usize,
    },

    /// List all topics and their partition counts.
    Metadata,
}

// ── Response ──────────────────────────────────────────────────────────────────

#[derive(Debug, Serialize, Deserialize, PartialEq)]
#[serde(tag = "type", rename_all = "snake_case")]
pub enum Response {
    TopicCreated { topic: String, partitions: u32 },
    Produced { partition: u32, offset: u64 },
    Record { offset: u64, payload: String },
    End,
    Metadata { topics: Vec<TopicMeta> },
    Error { message: String },
}

#[derive(Debug, Serialize, Deserialize, PartialEq)]
pub struct TopicMeta {
    pub name: String,
    pub partitions: u32,
}
